ls: Implement directory filter, optionally subfolders

mholt opened the pull request:
What is the purpose of this change? What does it change?
--------------------------------------------------------

Changes `restic ls` so that the syntax is:

```
restic ls snapshotID [dirs...]
```

instead of multiple snapshot IDs. Adds a new flag, `--recursive`, to allow matching subfolders of the given directories.

A directory filter constrains the walk to just the folder(s) specified in the arguments.

An `ls` that used to take 5 minutes now takes 50 seconds. (Most of that time is connecting to the repo, downloading the index, and decrypting it. The actual scan is instantaneous. It also uses less memory, since there are way fewer results.)

Was the change discussed in an issue or in the forum before?
------------------------------------------------------------

#1940

Checklist
---------

- [x] I have read the [Contribution Guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches)
- [ ] I have added tests for all changes in this PR
- [x] I have added documentation for the changes (in the manual)
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (template [here](https://github.com/restic/restic/blob/master/changelog/TEMPLATE))
- [x] I have run `gofmt` on the code in all commits
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits)
- [x] I'm done, this Pull Request is ready for review

mholt committed:
ls: Implement directory filter, optionally subfolders

mholt commented:
Dang, my commit from my other pull request snuck into this one. I'll try to fix that up later.

mholt committed:
Minor fixes/tweaks; add docs

fd0 commented on cmd/restic/cmd_ls.go:
```diff
@@ -59,19 +71,55 @@ func runLs(opts LsOptions, gopts GlobalOptions, args []string) error {
-	for sn := range FindFilteredSnapshots(ctx, repo, opts.Host, opts.Tags, opts.Paths, args) {
-		Verbosef("snapshot %s of %v at %s):\n", sn.ID().Str(), sn.Paths, sn.Time)
+	for sn := range FindFilteredSnapshots(ctx, repo, opts.Host, opts.Tags, opts.Paths, args[:1]) {
+		dirsToPrint := opts.Paths
```
Hm, I think the message should always print the source paths the snapshot was taken of. This is something completely different from the path within a snapshot. For example, consider a snapshot created with the following command:
```
$ cd /home/user
$ restic backup work
```

Then there will be a single path in the snapshot called `/home/user/work`, but the top-level directory in the snapshot will be `/work`. That's confusing for users, I think.

mholt replied:
Alright -- I can understand that. Reverting this change!

fd0 commented on cmd/restic/cmd_ls.go:
```diff
@@ -59,19 +71,55 @@ func runLs(opts LsOptions, gopts GlobalOptions, args []string) error {
+				var nodeDir string
+				if !opts.Recursive {
+					// only needed for exact directory match; i.e. no subfolders
+					nodeDir = filepath.Dir(nodepath)
```
`nodpath` is always slash-separated, as it is created by `path.Join()` (not `filepath.Join`), see https://godoc.org/github.com/restic/restic/internal/walker#WalkFunc

So we need to use `path.Dir()` here to make it work properly on Windows.

mholt replied:
Sounds good.

fd0 commented on cmd/restic/cmd_ls.go:
```diff
@@ -11,12 +13,21 @@ import (
+
+File listings can optionally be filtered by directories. Any
+positional arguments after the snapshot ID are interpreted as
+directory paths, and only files inside those directories will
```
I think it'd be good to have a hint here that `dir` needs to be slash-separated, and optionally starting with a `/`. How do you want to handle that? The path passed to the function in `walker.Walk()` will be slash-separated and starting with a `/` (so it's an absolute path)...

mholt replied:
Hmm, what do you mean _optionally_ starting with a `/`? I can mention it, I just don't quite understand what we need to communicate here yet.

fd0 replied:
Ah, sorry, I should have been clearer: the filter functions you wrote only work when the paths passed to `restic ls` are absolute (starting with a `/`), because the `nodepath` passed to the function in `walker.Walk()` starts with a slash. So I think we should document it.

Another option would be to just prefix the dirs with a slash if it isn't there already. That's what I meant with "optionally". Thinking about it, I'd suggest just requiring the paths to be absolute.

fd0 commented on cmd/restic/cmd_ls.go:
```diff
@@ -59,19 +71,55 @@ func runLs(opts LsOptions, gopts GlobalOptions, args []string) error {
+				var match bool
+				for _, dir := range dirs {
+					if opts.Recursive {
+						if strings.HasPrefix(nodepath, dir) {
```
That's not enough, for a node with the path `/foo/bar/baz` it would also be printed if restic was called like this: `restic ls latest /foo/bar/b`. Instead, call `fs.HasPathPrefix(dir, nodpath)`, which does the right thing: https://godoc.org/github.com/restic/restic/internal/fs#HasPathPrefix

mholt replied:
Ah, good catch. Thanks. My habits from Caddy path handling (which is different in HTTP requests) is carrying over here.

fd0 commented on cmd/restic/cmd_ls.go:
```diff
@@ -59,19 +71,55 @@ func runLs(opts LsOptions, gopts GlobalOptions, args []string) error {
+					}
+				}
+				if !match {
+					return true, nil
```
Ah, this may be confusing in the docs: When the walker function returns `ignore=true`, the **node** won't be visited again, which is identified by the tree ID, which may be the same for different directories at different paths. For example, consider a backup of a couple of empty dirs, e.g. `/home/foo/dir1` and `/other/dir2`, and calling `restic ls latest /other`. When the walker runs the function for `/home/foo/dir1` and `ignore=true` is returned (because it is not to be listed), it will add the tree ID for the empty directory to its list of trees to ignore, and the walker function will not be called again.

Instead, return `ignore=false` and return `walker.SkipNode` as the `error` instead.

I'm sorry that it's so confusing, the interface was adapted from `filepath.Walk`, and later I added the option to not only ignore stuff based on the path (return `SkipNode`) but also based on the tree ID (return `ignore=true`). Sigh. If you have an idea on how to improve the docs, please let me know!

mholt replied:
Ohhh... interesting! Thanks for clarifying that for me.

I wonder, in hindsight, is my usage of these return values correct in these places?

- https://github.com/restic/restic/blob/abc923f69391ae55b0319c31ff2560083c874322/cmd/restic/cmd_stats.go#L177-L184 (I noticed that in cmd_ls.go, you return false instead of true in these cases.)
- https://github.com/restic/restic/blob/abc923f69391ae55b0319c31ff2560083c874322/cmd/restic/cmd_stats.go#L236 (now I'm not sure if I didn't want to visit the node again or if I wanted to not visit the tree again)

Update: However, when I change this line to `return false, walker.SkipNode`, the tree is not walked at all. The ls does in fact work correctly with the line as-is... 🤔

fd0 replied:
I'm not sure what you mean by "this line", can you clarify?

fd0 replied:
FYI, the walker has many tests for all cases, maybe that can clarify it a bit, see:

https://github.com/restic/restic/blob/abc923f69391ae55b0319c31ff2560083c874322/internal/walker/walker_test.go#L216-L224

fd0 replied:
It doesn't, have a look at this repo: https://fd0.me/tmp/repo.tar.gz (the password is `foo`):

```
$ ./restic ls -l latest
repository 1271b4d6 opened successfully, password is correct
found 1 old cache directories in /home/fd0/.cache/restic, pass --cleanup-cache to remove them
snapshot 62bbf4fb of [] at 2018-08-11 22:09:31.33743769 +0200 CEST):
drwxr-xr-x  1000   100      0 2018-08-11 22:01:11 /test
drwxr-xr-x  1000   100      0 2018-08-11 22:02:58 /test/bar
-rw-r--r--  1000   100      0 2018-08-11 22:02:58 /test/bar/test.txt
drwxr-xr-x  1000   100      0 2018-08-11 22:04:19 /test/foo
-rw-r--r--  1000   100      0 2018-08-11 22:02:58 /test/foo/test.txt


$ ./restic ls -l latest /test/bar
repository 1271b4d6 opened successfully, password is correct
found 1 old cache directories in /home/fd0/.cache/restic, pass --cleanup-cache to remove them
snapshot 62bbf4fb of [/test/bar] at 2018-08-11 22:09:31.33743769 +0200 CEST):
-rw-r--r--  1000   100      0 2018-08-11 22:02:58 /test/bar/test.txt

$ ./restic ls -l latest /test/foo
repository 1271b4d6 opened successfully, password is correct
found 1 old cache directories in /home/fd0/.cache/restic, pass --cleanup-cache to remove them
snapshot 62bbf4fb of [/test/foo] at 2018-08-11 22:09:31.33743769 +0200 CEST):
```

The third call of `restic ls` did not print anything, although it should have printed the contents of the `test/foo` directory. The reason is that `/test/foo` and `/test/bar` have the same tree, which is on the ignore list because when `Walk()` processed `/test/bar`, it did not match the filter.

By the way, I think the directory itself (e.g. `/test/bar`) should also be printed.

fd0 replied:
> The SkipNode value seems very aggressive in what it skips... I'm still working on it though.

The semantics are modeled after `filepath.Walk()`...

> What do you mean by this? Which directory exactly, and printed where? 

I meant that if I run `restic ls latest /test/bar` then I think it's good to have the directory node `/test/bar` itself also printed, like this:
```
repository 1271b4d6 opened successfully, password is correct
found 1 old cache directories in /home/fd0/.cache/restic, pass --cleanup-cache to remove them
snapshot 62bbf4fb of [/test/bar] at 2018-08-11 22:09:31.33743769 +0200 CEST):
drwxr-xr-x  1000   100      0 2018-08-11 22:02:58 /test/bar
-rw-r--r--  1000   100      0 2018-08-11 22:02:58 /test/bar/test.txt
```

What do you think?

fd0 reviewed (changes requested):
So, first review comments are in. I'm sorry that it's so confusing, but that's life, full of interesting corner cases ;)

mholt committed:
Apply feedback and use SkipNode

mholt commented:
The commit I just pushed addresses most of the feedback. The logic was very different from what I expected! Sorry for the confusion.

Two things:

1. If you choose a shallow filter, e.g. `/` or `/home`, the directory listing is still sorta slow because it traverses subfolders (but doesn't print them) instead of skipping them entirely. I'm still working on the right logic for that...

2. I think the only other point of feedback that I haven't yet addressed is whether to match the folder itself, when specified in a filter. My particular use case does not need this, since I want to see the _contents_ of the filtered folder, but I could see why including the folder would be a good idea, too. I guess it could always be ignored. I'll revise that shortly.

mholt committed:
Speed up nonrecursive queries; include exact filter match

mholt commented:
I think there's a bug in SkipNode, somehow?

When I return SkipNode for a node that has the `nodepath` of a file, it skips the rest of the files in that directory.

Is that expected?

mholt committed:
Fix bug where some folder listings were empty

fd0 commented:
> When I return SkipNode for a node that has the nodepath of a file, it skips the rest of the files in that directory.
> Is that expected?

Yes, because I copied the behavior from `filepath.Walk()`:

>  If the function returns SkipDir when invoked on a directory, Walk skips the directory's contents entirely. If the function returns SkipDir when invoked on a non-directory file, Walk skips the remaining files in the containing directory.

mholt commented:
Ahh, I misunderstood the second sentence in the docs. Totally my bad. After re-reading, I understand now. Thanks!

fd0 commented:
I don't like this behavior at all, but until somebody spends time to revisit it, we're stuck ;)

mholt commented:
No worries. I think I made it work out alright. It's fast and accurate according to my tests, so I look forward to when you have a chance to try it!

fd0 commented:
Sooo. I'm sorry, but I think I found two bugs, using the test repo I mentioned yesterday:
```
$ restic ls -l latest

drwxr-xr-x  1000   100      0 2018-08-11 22:01:11 /test
drwxr-xr-x  1000   100      0 2018-08-11 22:02:58 /test/bar
-rw-r--r--  1000   100      0 2018-08-11 22:02:58 /test/bar/test.txt
drwxr-xr-x  1000   100      0 2018-08-11 22:04:19 /test/foo
-rw-r--r--  1000   100      0 2018-08-11 22:02:58 /test/foo/test.txt
```

This is empty:
```
$ restic ls -l /test/foo
```

Since I could not hold my fingers still, I've already debugged the issues and will push some commits soon.

fd0 commented on cmd/restic/cmd_ls.go:
```diff
@@ -59,20 +74,89 @@ func runLs(opts LsOptions, gopts GlobalOptions, args []string) error {
 		return err
 	}
 
+	// extract any specific directories to walk
```
I'd do that before opening the repo, so users get a feedback on the dirs they specified (e.g. when they aren't absolute) early on, before any network traffic.

fd0 commented on cmd/restic/cmd_ls.go:
```diff
@@ -59,20 +74,89 @@ func runLs(opts LsOptions, gopts GlobalOptions, args []string) error {
+				// are not in matching trees or will not lead us to matching trees
+				var walk bool
+				for _, dir := range dirs {
+					approachingMatchingTree := fs.HasPathPrefix(nodeDir, dir)
```
This needs to use `nodepath` instead of `nodeDir`, otherwise one level is walked too much. 

When the nodepath is `/test/foo` and the filter dir is `/test/bar`, then this would still evaluate:
```go
fs.HasPathPrefix("/test", "/test/bar")
```

It should do the following:
```go
fs.HasPathPrefix("/test/foo", "/test/bar")
```

fd0 commented on cmd/restic/cmd_ls.go:
```diff
@@ -59,20 +74,89 @@ func runLs(opts LsOptions, gopts GlobalOptions, args []string) error {
+					}
+				}
+				if !walk {
+					return false, walker.SkipNode
```
We need to be careful here: When we return `SkipNode` for a node that is not a dir (e.g. a file or other type), then the remaining nodes in the same tree are skipped!

fd0 committed:
ls: Check dirs before opening the repository

Users get feedback instantly, and before any expensive network calls
have been made.

fd0 committed:
ls: Use nodepath for filter

fd0 committed:
ls: Only skip directory nodes

Special case for Walk(): When SkipDir is returned for a non-dir node,
the remaining nodes for the current tree are skipped. We don't want
that.

fd0 committed:
ls: Add comments

fd0 committed:
ls: Rework and simplify logic

This commit introduces two functions: withinDir() and
approachingMatchingTree()

Both bind the list of directories with a closure, so we don't need to
iterate over the list in the function passed to Walk(). This reduces the
indentation level and since we can just use return, we don't need the
breaks any more.

The case that len(dirs) == 0 can also be handled by the functions with a
return, which saves another indentation level.

The main function body of the function passed to Walk() was reduced to
three cases:

 * Within one of the dirs: Print the node, and if recursive operation is
   requested, directly return, so the walker continues recursive
   traversal

 * Approaching one of the dirs: don't print anything, but continue
   recursive traversal.

 * Nothing of the two: abort walking this branch of the tree.

mholt commented:
😮 You are a _saint_. Thanks for your patience and initiative here. I know sometimes contributions can be a burden too -- sorry if this is one of those times!

I've pulled your changes and inspected the code. That is definitely cleaner than my approach, and I see now where the main bug was in mine.

Your changes work for me in all my tests, and it's pretty speedy, too. Yours even works if there's a trailing slash in the filter :) The --recursive flag and multiple filters all seem to work as expected.

I agree tests would be good, although I am not sure where to add tests for the ls command, nor quite sure a strategy for how to test it.

Thank you again! :+1:

fd0 commented:
> I know sometimes contributions can be a burden too -- sorry if this is one of those times!

Don't worry, it was not :)
